src/commands/githubActions/bumpUpstream/index.ts

```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
-  const repoData = await thisRepo.getRepo();
-  const remoteBranchExists = await thisRepo.branchExists(branch);
-  const localBranchExists = await getLocalBranchExists(branch);
-  const branchExists = remoteBranchExists || localBranchExists;
-
-  if (branchExists) {
-    console.log(`Branch ${branch} already exists`);
-    return;
-    // TODO, do some checking
-  }
+  return upstreamRepoVersions;
+}
 
-  // index by repoSlug, must be unique
-  const versionsToUpdateMap = new Map<string, VersionToUpdate>();
-
-  for (const [serviceName, service] of Object.entries(compose.services))
-    if (typeof service.build === "object" && service.build.args)
-      for (const [argName, argValue] of Object.entries(service.build.args)) {
-        const upstreamRepoVersion = upstreamRepoVersions.get(argName);
-        if (!upstreamRepoVersion) continue;
-
-        const currentVersion = argValue;
-        const { repoSlug, newVersion } = upstreamRepoVersion;
-        if (currentVersion === newVersion) continue;
-
-        // Update current version
-        compose.services[serviceName].build = {
-          ...service.build,
-          args: {
-            ...service.build.args,
-            [argName]: newVersion
-          }
-        };
-        // Use a Map since there may be multiple matches for the same argName
-        versionsToUpdateMap.set(repoSlug, {
-          repoSlug,
-          newVersion,
-          currentVersion
-        });
-      }
+function updateComposeVersions(dir: string, compose: Compose, upstreamRepoVersions: UpstreamRepoMap): ComposeVersionsToUpdate {
+  const newCompose = JSON.parse(JSON.stringify(compose)); // Deep copy
+  const versionsToUpdate: ComposeVersionsToUpdate = {};
 
-  if (versionsToUpdateMap.size == 0) {
-    console.log("All versions are up-to-date");
-    return;
+  for (const [serviceName, service] of Object.entries(compose.services)) {
+
+    if (typeof service.build !== "object" || !service.build.args)
+      continue;
+
+    for (const [argName, currentVersion] of Object.entries(service.build.args)) {
+      const upstreamVersionInfo = upstreamRepoVersions[argName];
+
+      if (!upstreamVersionInfo || currentVersion === upstreamVersionInfo.newVersion)
+        continue;
+
+      newCompose.services[serviceName].build.args[argName] = upstreamVersionInfo.newVersion;
+
+      versionsToUpdate[upstreamVersionInfo.repoSlug] = {
+        newVersion: upstreamVersionInfo.newVersion,
+        currentVersion,
+      };
+    }
   }
 
-  const versionsToUpdate = Array.from(versionsToUpdateMap.values());
-  manifest.upstreamVersion = getUpstreamVersionTag(versionsToUpdate);
+  if (!isEmpty(versionsToUpdate))
+    writeCompose(newCompose, { dir });
 
-  const ethProviderAvailable = await getFirstAvailableEthProvider({
-    providers: ethProviders
-  });
+  return versionsToUpdate;
+}
 
-  if (!ethProviderAvailable)
-    throw Error(`No eth provider available. Tried: ${ethProviders.join(", ")}`);
+async function updateManifest({
```

Comment:
This function has a lot of logic embedded, could you add a detailed description of what it does, what it takes and what it returns

Reply:
Okay!